File: julia/2021/day2.jl
include("util.jl")

# read relative to this dir
cd(@__DIR__)

function parseInput(s)
    v = split(s, " ")
    [v[1], parse(Int, v[2])]
end

input = readlines("input/day2.txt")
parsed = [parseInput(s) for s in input]
parsed

## part 1

function part1()
    x = 0
    y = 0

    for line in parsed
        d = line[1]
        n = line[2]
        if d == "forward"
            x += n
        elseif d == "down"
            y += n
        elseif d == "up"
            y -= n
        end
    end
    println(x * y)
end
part1()

## part 2

function part2()
    x = 0
    y = 0
    aim = 0

    for line in parsed
        d = line[1]
        n = line[2]
        if d == "forward"
            x += n
            y += aim * n
        elseif d == "down"
            aim += n
        elseif d == "up"
            aim -= n
        end
    end
    println(x * y)
end
part2()

File: typescript/src/day2.ts
import { readLines } from './util'
import * as _ from 'lodash';

async function main(): Promise<void> {
  const input = await readLines("input/day2.txt")

  await part1(input)
  await part2(input)
}

async function part1(input: string[]): Promise<void> {
  let s = 0
  for (const line of input) {
    let [m, e] = line.split(" ")

    let shape = {
      'X': 1,
      'Y': 2,
      'Z': 3
    }[e]

    let win = {
      'A X': 3,
      'A Y': 6,
      'A Z': 0,
      'B X': 0,
      'B Y': 3,
      'B Z': 6,
      'C X': 6,
      'C Y': 0,
      'C Z': 3,
    }[line]

    let score = shape + win
    s += score
  }
  console.log(s)
}

async function part2(input: string[]): Promise<void> {
  let s = 0
  for (const line of input) {
    let [m, e] = line.split(" ")

    let base = {
      'A': 0,
      'B': 1,
      'C': 2
    }[m]

    let offset = {
      'X': 2,
      'Y': 0,
      'Z': 1
    }[e]

    let win = {
      'X': 0,
      'Y': 3,
      'Z': 6
    }[e]

    let shape = 1 + (base + offset) % 3

    let score = shape + win
    s += score
  }
  console.log(s)
}

main()

// Make sure this is detected as a module and not compiled in the global scope
export default null